pkg/codegen/schema/schema_test.go

```diff
@@ -2039,3 +2039,108 @@ func TestRoundtripAliasesYAML(t *testing.T) {
 
 	assert.YAMLEq(t, string(schemaBytes), string(yamlData))
 }
+
+func TestFunctionToken(t *testing.T) {
+	tests := []struct {
+		name     string
+		token    string
+		expected hcl.Diagnostics
+	}{
+		{
+			name:     "valid_token_for_provider_method",
+			token:    "pulumi:providers:terraform-provider/providerMethod",
+			expected: hcl.Diagnostics(nil),
+		},
+		{
+			name:     "valid_token_without_hyphens",
+			token:    "test:index:getFunction",
+			expected: hcl.Diagnostics(nil),
+		},
+		{
+			name:     "valid_token_with_hyphens",
+			token:    "test:index:get-function-data",
+			expected: hcl.Diagnostics(nil),
+		},
+		{
+			name:     "valid_token_with_leading_hyphen",
+			token:    "test:index:-getFunction",
+			expected: hcl.Diagnostics(nil),
+		},
+		{
+			name:     "valid_token_with_consecutive_hyphens",
+			token:    "test:index:get--function",
+			expected: hcl.Diagnostics(nil),
+		},
+		{
+			name:  "invalid_token_with_invalid_package",
+			token: "123:index:getFunction",
+			expected: hcl.Diagnostics{
+				{
+					Severity: hcl.DiagError,
+					Summary:  "#/functions/123:index:getFunction: doesn't validate with '/$defs/functionToken'",
+					Detail:   "",
+				},
+				{
+					Severity: hcl.DiagError,
+					Summary:  "#/functions/123:index:getFunction: does not match pattern '^[a-zA-Z][-a-zA-Z0-9_]*:([^0-9][a-zA-Z0-9._/-]*)?:[^0-9][a-zA-Z0-9._/-]*$'",
+					Detail:   "",
+				},
+				{
+					Severity: hcl.DiagError,
+					Summary:  "#/functions/123:index:getFunction: invalid token '123:index:getFunction' (must have package name 'test')",
+					Detail:   "",
+				},
+			},
+		},
+		{
+			name:  "invalid_token_with_invalid_module",
+			token: "test:123:getFunction",
+			expected: hcl.Diagnostics{
+				{
+					Severity: hcl.DiagError,
+					Summary:  "#/functions/test:123:getFunction: doesn't validate with '/$defs/functionToken'",
+					Detail:   "",
+				},
+				{
+					Severity: hcl.DiagError,
+					Summary:  "#/functions/test:123:getFunction: does not match pattern '^[a-zA-Z][-a-zA-Z0-9_]*:([^0-9][a-zA-Z0-9._/-]*)?:[^0-9][a-zA-Z0-9._/-]*$'",
+					Detail:   "",
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
```suggestion
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
```
this is what the linter is after I think. Same for the other cases